File: Recursos/data.structures/haskell/DataStructures/Graph/EulerianCycle.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.EulerianCycle(isEulerian, eulerianCycle) where

import DataStructures.Graph.Graph
import Data.List

--H.1)
isEulerian :: Eq a => Graph a -> Bool
isEulerian g = null [ v | v <- vertices g, (degree g v) `mod` 2 /= 0]

-- H.2)
remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove g (v,u) = g'''
      where
          g' = deleteEdge g (v,u)
          g'' = if (degree g' v == 0) then deleteVertex g' v else g'
          g''' = if (degree g'' u == 0) then deleteVertex g'' u else g''

-- H.3)
extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle g v0 = aux g v0 [v0]
    where
        aux g v xs
          | u == v0 = (g', u:xs)
          | otherwise = aux g' u  (u:xs)
          where
              u = head (successors g v)
              g'= remove g (v,u)

-- H.4)
connectCycles :: (Eq a) => Path a -> Path a -> Path a
connectCycles [] (y:ys) = y:ys
connectCycles (x:xs) (y:ys)
    | x == y = y:ys ++ xs
    | otherwise = x:(connectCycles xs (y:ys))

-- H.5)
vertexInCommon :: Eq a => Graph a -> Path a -> a
vertexInCommon g cycle = head [v | v <- vertices g, elem v cycle]

-- H.6)
eulerianCycle :: Eq a => Graph a -> Path a
eulerianCycle g
  | isEulerian g = aux g (head (vertices g)) []
  | otherwise = []
    where
      aux g v xs
        | isEmpty g = xs
        | otherwise = aux g' w xs'
        where
          (g', cycle) = extractCycle g v
          xs' = connectCycles xs cycle
          w = vertexInCommon g' xs'

File: Exámenes/2016/Febrero/Java/eulerianCycle/src/dataStructures/graph/EulerianCycle.java
/**
 * Student's name:
 * Student's group:
 *
 * Data Structures. Grado en Informática. UMA.
 */

package dataStructures.graph;

import dataStructures.list.*;

public class EulerianCycle<V> {
    private List<V> eCycle;

    @SuppressWarnings("unchecked")
    public EulerianCycle(Graph<V> g) {
        Graph<V> graph = (Graph<V>) g.clone();
        eCycle = eulerianCycle(graph);
    }

    public boolean isEulerian() {
        return eCycle != null;
    }

    public List<V> eulerianCycle() {
        return eCycle;
    }

    // J.1
    private static <V> boolean isEulerian(Graph<V> g) {
        boolean res = true;

        for (V v : g.vertices()) {
            res = res && (g.degree(v) % 2 == 0);
        }

        return res;
    }

    // J.2
    private static <V> void remove(Graph<V> g, V v, V u) {
        g.deleteEdge(v, u);

        for (V ver : g.vertices()) {
            if (g.degree(ver) == 0) {
                g.deleteVertex(ver);
            }
        }
    }

    // J.3
    private static <V> List<V> extractCycle(Graph<V> g, V v0) {
        List<V> ciclo = new LinkedList<>();
        V v = v0, u = g.successors(v).iterator().next();

        ciclo.append(v);

        while (v0 != u) {
            remove(g, v, u);
            ciclo.append(u);

            v = u;
            u = g.successors(v).iterator().hasNext() ? g.successors(v).iterator().next() : v0;
        }

        ciclo.append(v0);

        return ciclo;
    }

    // J.4
    private static <V> void connectCycles(List<V> xs, List<V> ys) {
        if (xs.isEmpty()) {
            for (V v : ys) {
                xs.append(v);
            }

        } else {
            int i = 0;

            while (xs.get(i) != ys.get(0)) {
                i++;
            }

            for (V v : ys) {
                i++;
                xs.insert(i, v);

            }
        }
    }

    // J.5
    private static <V> V vertexInCommon(Graph<V> g, List<V> cycle) {
        int i = 0;
        V res = null;

        if (cycle.isEmpty()) {
            res = g.vertices().iterator().next();

        } else{
            while (res == null) {    // Asumiendo que siempre hay al menos uno
                V actual = cycle.get(i);
                g.vertices().isElem(actual);

                res = actual;
            }
        }

        return res;
    }

    // J.6
    private static <V> List<V> eulerianCycle(Graph<V> g) {
        List<V> ciclo;

        if (!isEulerian(g)) {
            ciclo = null;       // El test interpreta un ciclo nulo como muetra de no-euleriano

        } else {
            ciclo = new LinkedList<>();

            while (!g.isEmpty()) {
                connectCycles(ciclo, extractCycle(g,vertexInCommon(g, ciclo)));
            }
        }

        return ciclo;
    }
}
